File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/StartGame.java
package org.emulinker.kaillera.controller.v086.protocol;

import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.util.UnsignedUtil;

public abstract class StartGame extends V086Message {
  public static final byte ID = 0x11;

  public abstract int val1();

  public abstract short playerNumber();

  public abstract short numPlayers();

  @Override
  public int getBodyLength() {
    return 5;
  }

  @Override
  public void writeBodyTo(ByteBuffer buffer) {
    buffer.put((byte) 0x00);
    UnsignedUtil.putUnsignedShort(buffer, val1());
    UnsignedUtil.putUnsignedByte(buffer, playerNumber());
    UnsignedUtil.putUnsignedByte(buffer, numPlayers());
  }

  public static StartGame parse(int messageNumber, ByteBuffer buffer)
      throws ParseException, MessageFormatException {
    if (buffer.remaining() < 5) throw new ParseException("Failed byte count validation!");

    byte b = buffer.get();

    if (b != 0x00) throw new ParseException("Failed byte count validation!");

    int val1 = UnsignedUtil.getUnsignedShort(buffer);
    short playerNumber = UnsignedUtil.getUnsignedByte(buffer);
    short numPlayers = UnsignedUtil.getUnsignedByte(buffer);

    if (val1 == 0xFFFF && playerNumber == 0xFF && numPlayers == 0xFF)
      return StartGame_Request.create(messageNumber);
    else return StartGame_Notification.create(messageNumber, val1, playerNumber, numPlayers);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/StartGame.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.util.UnsignedUtil

abstract class StartGame : V086Message() {
  override val messageId = ID

  abstract val val1: Int
  abstract val playerNumber: Short
  abstract val numPlayers: Short

  override val bodyLength: Int
    get() = 5

  public override fun writeBodyTo(buffer: ByteBuffer) {
    buffer.put(0x00.toByte())
    UnsignedUtil.putUnsignedShort(buffer, val1)
    UnsignedUtil.putUnsignedByte(buffer, playerNumber.toInt())
    UnsignedUtil.putUnsignedByte(buffer, numPlayers.toInt())
  }

  companion object {
    const val ID: Byte = 0x11
    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): StartGame {
      if (buffer.remaining() < 5) throw ParseException("Failed byte count validation!")
      val b = buffer.get()
      if (b.toInt() != 0x00) throw ParseException("Failed byte count validation!")
      val val1 = UnsignedUtil.getUnsignedShort(buffer)
      val playerNumber = UnsignedUtil.getUnsignedByte(buffer)
      val numPlayers = UnsignedUtil.getUnsignedByte(buffer)
      return if (val1 == 0xFFFF && playerNumber.toInt() == 0xFF && numPlayers.toInt() == 0xFF)
          StartGame_Request(messageNumber)
      else StartGame_Notification(messageNumber, val1, playerNumber, numPlayers)
    }
  }
}
